### PR TITLE
[MultiThreading] Add Image plugin Data types in DataExchange component

### PR DESCRIPTION
**Pull request by agpetit**
This PR gives the ability to exchange images between different threads in the MultiThreading plugin. It would be especially usefull for multithreading a camera grabber component and other components (for instance computer vision algorithms). In more details: 
- It consists in adding,  in the DataExchange component, exchanges of image data types (ImageUC, ImageUS, ImageF). It involves a depedency with the image plugin, handled in the CMakeLists through a preprocessor directive (WITH_IMAGE) defined in config.h.in.
- A simple example passing an image loaded from an ImageContainer component (thread 1) to an ImageViewer component (thread 2), to display the image in the GUI, is added.
- An issue with OpenGL in the draw() function in the Kinect component in the image plugin has also been fixed, by using the SOFA_NO_OPENGL directive (in the CMakeLists and in Kinect.h). Indeed, with the lib FREENECT detected, the OpenGL functions in draw() may not be compiled given the OpenGL support of SOFA.    

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by fspadoni** [comment]
That's good to support new data types in the DataExchange component and provide an example scene. Thanks.

File: applications/plugins/image/Kinect.h [hunk 1]
```diff
@@ -502,6 +502,7 @@ class Kinect : public virtual core::objectmodel::BaseObject
 
     void draw(const core::visual::VisualParams* vparams)
     {
+#ifndef SOFA_NO_OPENGL
```
Comment (epernod) [inline applications/plugins/image/Kinect.h:505]: just for my knowledge, why is this needed?
Is it linked with the multi-thread use?

**Comment by agpetit** [comment]
@fspadoni i am using this feature for multithreading a RealSense RGB-D camera, through a RealSense grabber component in the image plugin (for which i will prepare another PR :-)), with other components in another plugin for non-rigid tracking goals. And it works well!

File: applications/plugins/image/Kinect.h (hunk 1 above)
Comment (agpetit) [reply applications/plugins/image/Kinect.h:505]: Hi @epernod !
No it is not directly linked to multihreading. The new feature with image data tyes in MultiThreading requires the image plugin. And compiling it I noticed this error with OpenGL in Kinect.h (when the lib FREENECT is found), so I just added this small fix, but perhaps I should have done it in another PR.

Comment (epernod) [reply applications/plugins/image/Kinect.h:505]: no, it was just for understanding.

**Comment by fspadoni** [comment]
@hugtalbot , I think it's not possible to move the declaration of the new DataExchange type class in the Image plugin.
The DataExchange component is a template class and it can't be used across the dynamic libraries but it must be instantiate and exported in the library where they are defined for every data type.
For instance in the image plugin the DataExchange template class can't be used to defined the class DataExchange< ImageUC > because the DataExchange template class is defined in another dynamic library. That should be allow only across static libraries.
correct me if I am wrong.
I didn't realize that this morning when we talked about it at the SOFA meeting.

**Comment by damienmarchal** [comment]
To inverse the dependency the registering should be moved into image.cpp
```cpp
#ifdef SOFA_WITH_MULTITHREADING

int DataExchangeClass = core::RegisterObject("DataExchange")
   .add< DataExchange< sofa::defaulttype::ImageUC> >()
   .add< DataExchange< sofa::defaulttype::ImageUS> >()
   .add< DataExchange< sofa::defaulttype::ImageF> >();

#endif // SOFA_WITH_MULTITHREADING
```

About the @fspadoni's answer I think that template can be instantiated as long as the template definition is #included (the .inl or the .h)

**Comment by epernod** [comment]
[ci-build][with-scene-tests]
